### PR TITLE
Add shard exp on fsdp custom mesh

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

This PR introduces a new custom mesh and rule that enabling FSDP sharding exp dimension for weights in MoE component. When non-elementwise optimization is enabled, e.g. Muon, this custom mesh show better benefits. 

Compared with previous implementation of `shard_exp_on_fsdp`, this custom mesh support mixture of FSDP and EP.

Use `custom_mesh_and_rule=shard-exp-on-fsdp` to enable.

# Tests

## Performance Regression

Tpu7x-8 on DSv2-16b, the losses of 10 steps perfectly match and performance are similar, see https://diff.googleplex.com/#key=r2vyDl7870kN.

## Support on FSDP + EP

* Support FSDP=4 and EP=2 (a2a): https://paste.googleplex.com/4617065838280704
* Support FSDP=4 and EP=2 (AG-RS): https://paste.googleplex.com/5176088446763008

## Support on explicit sharding

* Pure FSDP=8: https://paste.googleplex.com/4860389375475712
* FSDP=4 and EP=2: https://paste.googleplex.com/4890472467267584

## Performance improvement on Muon optimizer
* FSDP+EP MFU improved 120%: https://diff.googleplex.com/#key=ieiMesgH7AMi
* Pure FSDP same performance: https://diff.googleplex.com/#key=FdgAlFcTiTHF

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -1149,10 +1145,11 @@ def get_quantization_dtypes():
         rhs_quantize_dtype = quant_dg.fwd.dg_quantizer.rhs.numerics.get_dtype()
       return lhs_quantize_dtype, rhs_quantize_dtype
 
-    def gmm(inputs, kernel, tiling, group_sizes, expert_assignments, weight_gather_axes):
+    def gmm(inputs, kernel, tiling, group_sizes, expert_assignments, weight_gather_axes=None):
```
Comment (Shuwen-Fang) [inline src/maxtext/layers/moe.py]: can we do weight_gather_axes = []

File: src/maxtext/utils/sharding.py [hunk 2]
```diff
@@ -732,6 +732,31 @@ def _remove_fsdp_from_partition_spec(named_sharding):
   return jax.tree.map(_remove_fsdp_from_partition_spec, sharding_tree)
 
 
+def remove_fsdp_pspec(pspec):
+  """Removes 'fsdp' and 'fsdp_transpose' from a PartitionSpec."""
+  if isinstance(pspec, jax.sharding.PartitionSpec):
```
Comment (Shuwen-Fang) [inline src/maxtext/utils/sharding.py]: in what scenario is it not a `jax.sharding.PartitionSpec` type?

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -1251,25 +1241,13 @@ def get_routed_moe_shardings(is_batch_sharded_by_expert):
...
-      else:
-        # These are the main shardings used by default - they use funky rules to AG over FSDP.
-        w0_pspec = self._logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-        w1_pspec = self._logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-        wo_pspec = self._logical_to_mesh_axes(("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+      w0_pspec = self._logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
+      w1_pspec = self._logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
+      wo_pspec = self._logical_to_mesh_axes(("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+      # Update kernel pspec for FSDP AG
+      w0_pspec = remove_fsdp_pspec(w0_pspec)
```
Comment (Shuwen-Fang) [inline src/maxtext/layers/moe.py]: For my own understanding, why remove fsdp from pspec?

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (NuojCheng) [reply src/maxtext/layers/moe.py]: Tried it first but got some pylint warning..

File: src/maxtext/utils/sharding.py (hunk 2 above)
Comment (NuojCheng) [reply src/maxtext/utils/sharding.py]: in some cases it might be none. Since it is a shared funciton in sharding.py, I tried to make this funciton as general-purpose as possible.

File: src/maxtext/layers/moe.py (hunk 3 above)
Comment (NuojCheng) [reply src/maxtext/layers/moe.py]: For the sparse matmul wrapper function, we FSDP all gather weights before starting the shard map. I think it is just a decision made earlier.

File: src/maxtext/utils/sharding.py (hunk 2 above)
Comment (Shuwen-Fang) [reply src/maxtext/utils/sharding.py]: nit, can we change it to: 

if pspec == None: 
  return psepc
new_spec = []
...

**Review by Shuwen-Fang — APPROVED** [review]

File: src/maxtext/utils/sharding.py (hunk 2 above)
Comment (NuojCheng) [reply src/maxtext/utils/sharding.py]: good idea! done

**Review by huytransformer — APPROVED** [review]
